### PR TITLE
[Merged by Bors] - Move WaitPeers utlity to p2p/peers

### LINKED ISSUE #2640: Move code to wait for peers from peersync module to p2p
## Description

This code https://github.com/spacemeshos/go-spacemesh/blob/develop/timesync/peersync/peers.go belongs in p2p module. Probably generis Peers listener can be extended with an additional funtion.

### LINKED ISSUE #2648: Occasional panic on shutdown
## Description

I shut down a node running locally and got a panic while shutting down. After restarting, I wasn't able to reproduce it

## Steps To Reproduce
1. Start a node
2. Wait some time
3. Press Ctrl-C
4. The error may appear

## Expected Behavior

No panic

## Environment
Please complete the following information:
 - OS: 
 
macOS 11
 - Node Version: 
 
`v0.0.0-unreleased+33c175d4cbae624b4c9de9b5879fec781f305fa0`
 - App Version (if applicable): 
 
N/A

## Additional Resources 

```
2021-08-09T18:31:10.490+0300	INFO	d1599.fetcher      	remote peer requested non existing hash	{"node_id": "d1599f8566555276993fb7bb2f078d2b87d8e1c3879ddf26d386ba3c96b1e8a2", "module": "fetcher", "sessionId": "6810e1d0-7f2d-4e16-9d56-f7b26883fdc9", "hash": "ba3c42a73d", "hint": "IVDB", "errmsg": "leveldb: not found", "name": "fetcher"}
2021-08-09T18:31:10.676+0300	INFO	d1599.fetcher      	remote peer requested non existing hash	{"node_id": "d1599f8566555276993fb7bb2f078d2b87d8e1c3879ddf26d386ba3c96b1e8a2", "module": "fetcher", "sessionId": "6810e1d0-7f2d-4e16-9d56-f7b26883fdc9", "hash": "ba3c42a73d", "hint": "IVDB", "errmsg": "leveldb: not found", "name": "fetcher"}
2021-08-09T18:31:10.740+0300	INFO	d1599.fetcher      	remote peer requested non existing hash	{"node_id": "d1599f8566555276993fb7bb2f078d2b87d8e1c3879ddf26d386ba3c96b1e8a2", "module": "fetcher", "sessionId": "6810e1d0-7f2d-4e16-9d56-f7b26883fdc9", "hash": "ba3c42a73d", "hint": "IVDB", "errmsg": "leveldb: not found", "name": "fetcher"}
^C2021-08-09T18:31:11.179+0300	INFO	d1599.app          	Received an interrupt, stopping services...
	{"node_id": "d1599f8566555276993fb7bb2f078d2b87d8e1c3879ddf26d386ba3c96b1e8a2", "module": "app"}
2021-08-09T18:31:11.179+0300	INFO	d1599.p2p          	shutting down protocol	{"node_id": "d1599f8566555276993fb7bb2f078d2b87d8e1c3879ddf26d386ba3c96b1e8a2", "module": "p2p", "p2pid": "83VdmNkSgdiJ63DqPyojr4Gu5ZUJ4SMn3a192U3LoM8D", "protocol": "/udp/v2/discovery", "name": "p2p"}
2021-08-09T18:31:11.179+0300	INFO	d1599.layerFetcher 	shutting down protocol	{"node_id": "d1599f8566555276993fb7bb2f078d2b87d8e1c3879ddf26d386ba3c96b1e8a2", "module": "layerFetcher", "protocol": "/layers/2.0/", "name": "layerFetcher"}
2021-08-09T18:31:11.179+0300	ERROR	d1599.p2p          	propagate event loop stopped: protocol shutdown	{"node_id": "d1599f8566555276993fb7bb2f078d2b87d8e1c3879ddf26d386ba3c96b1e8a2", "module": "p2p", "p2pid": "83VdmNkSgdiJ63DqPyojr4Gu5ZUJ4SMn3a192U3LoM8D", "sessionId": "06df3fee-daef-4b3c-97ce-b22ba7ea9839"}
2021-08-09T18:31:11.179+0300	INFO	d1599.fetcher      	shutting down protocol	{"node_id": "d1599f8566555276993fb7bb2f078d2b87d8e1c3879ddf26d386ba3c96b1e8a2", "module": "fetcher", "protocol": "/sync/2.0/", "name": "fetcher"}
2021-08-09T18:31:11.179+0300	ERROR	d1599.p2p          	addrbook refresh error	{"node_id": "d1599f8566555276993fb7bb2f078d2b87d8e1c3879ddf26d386ba3c96b1e8a2", "module": "p2p", "p2pid": "83VdmNkSgdiJ63DqPyojr4Gu5ZUJ4SMn3a192U3LoM8D", "errmsg": "bootstrap canceled by signal", "name": "p2p"}
2021-08-09T18:31:11.179+0300	WARN	d1599.p2p          	needs 6 more peers	{"node_id": "d1599f8566555276993fb7bb2f078d2b87d8e1c3879ddf26d386ba3c96b1e8a2", "module": "p2p", "p2pid": "83VdmNkSgdiJ63DqPyojr4Gu5ZUJ4SMn3a192U3LoM8D"}
2021-08-09T18:31:11.179+0300	INFO	d1599.p2p          	priority queue was closed, exiting	{"node_id": "d1599f8566555276993fb7bb2f078d2b87d8e1c3879ddf26d386ba3c96b1e8a2", "module": "p2p", "p2pid": "83VdmNkSgdiJ63DqPyojr4Gu5ZUJ4SMn3a192U3LoM8D", "sessionId": "06df3fee-daef-4b3c-97ce-b22ba7ea9839", "errmsg": "the queue is closed", "name": "p2p"}
2021-08-09T18:31:11.180+0300	INFO	00000.defaultLogger	app cleanup starting...
2021-08-09T18:31:11.180+0300	INFO	00000.defaultLogger	stopping grpc service
2021-08-09T18:31:11.181+0300	INFO	00000.defaultLogger	Stopping new grpc server...
2021-08-09T18:31:11.181+0300	INFO	d1599.app          	closing block producer	{"node_id": "d1599f8566555276993fb7bb2f078d2b87d8e1c3879ddf26d386ba3c96b1e8a2", "module": "app"}
2021-08-09T18:31:11.183+0300	INFO	d1599.blockBuilder 	database closed	{"node_id": "d1599f8566555276993fb7bb2f078d2b87d8e1c3879ddf26d386ba3c96b1e8a2", "module": "blockBuilder", "file": "/Users/nkryuchkov/spacemesh/20/builder", "name": "blockBuilder"}
2021-08-09T18:31:11.183+0300	INFO	d1599.app          	closing clock	{"node_id": "d1599f8566555276993fb7bb2f078d2b87d8e1c3879ddf26d386ba3c96b1e8a2", "module": "app"}
2021-08-09T18:31:11.183+0300	INFO	d1599.clock        	closing clock 0xc0000be900	{"node_id": "d1599f8566555276993fb7bb2f078d2b87d8e1c3879ddf26d386ba3c96b1e8a2"}
2021-08-09T18:31:11.183+0300	INFO	d1599.clock        	closed clock 0xc0000be900	{"node_id": "d1599f8566555276993fb7bb2f078d2b87d8e1c3879ddf26d386ba3c96b1e8a2"}
2021-08-09T18:31:11.183+0300	INFO	d1599.app          	Stopping tortoise beacon...	{"node_id": "d1599f8566555276993fb7bb2f078d2b87d8e1c3879ddf26d386ba3c96b1e8a2", "module": "app"}
2021-08-09T18:31:11.183+0300	INFO	d1599.tBeacon      	Closing TORTOISE_BEACON_PROTOCOL	{"node_id": "d1599f8566555276993fb7bb2f078d2b87d8e1c3879ddf26d386ba3c96b1e8a2", "module": "tBeacon"}
2021-08-09T18:31:11.183+0300	INFO	d1599.app          	closing poet listener	{"node_id": "d1599f8566555276993fb7bb2f078d2b87d8e1c3879ddf26d386ba3c96b1e8a2", "module": "app"}
2021-08-09T18:31:11.183+0300	INFO	d1599.app          	closing atx builder	{"node_id": "d1599f8566555276993fb7bb2f078d2b87d8e1c3879ddf26d386ba3c96b1e8a2", "module": "app"}
2021-08-09T18:31:11.183+0300	INFO	d1599.app          	closing hare	{"node_id": "d1599f8566555276993fb7bb2f078d2b87d8e1c3879ddf26d386ba3c96b1e8a2", "module": "app"}
2021-08-09T18:31:11.184+0300	INFO	d1599.app          	closing p2p	{"node_id": "d1599f8566555276993fb7bb2f078d2b87d8e1c3879ddf26d386ba3c96b1e8a2", "module": "app"}
2021-08-09T18:31:11.184+0300	INFO	d1599.p2p          	Address manager shutting down	{"node_id": "d1599f8566555276993fb7bb2f078d2b87d8e1c3879ddf26d386ba3c96b1e8a2", "module": "p2p", "p2pid": "83VdmNkSgdiJ63DqPyojr4Gu5ZUJ4SMn3a192U3LoM8D"}
2021-08-09T18:31:11.184+0300	INFO	d1599.hare         	broker exiting	{"node_id": "d1599f8566555276993fb7bb2f078d2b87d8e1c3879ddf26d386ba3c96b1e8a2", "module": "hare"}
2021-08-09T18:31:11.184+0300	INFO	d1599.hare         	priority queue was closed, exiting	{"node_id": "d1599f8566555276993fb7bb2f078d2b87d8e1c3879ddf26d386ba3c96b1e8a2", "module": "hare", "protocol": "HARE_PROTOCOL_broker", "sessionId": "bffc2663-b381-423e-ae2b-c8f44b28958e", "latest_layer": 11, "errmsg": "the queue is closed", "name": "hare"}
2021-08-09T18:31:11.184+0300	INFO	d1599.poetListener 	listening stopped	{"node_id": "d1599f8566555276993fb7bb2f078d2b87d8e1c3879ddf26d386ba3c96b1e8a2", "module": "poetListener"}
2021-08-09T18:31:11.189+0300	ERROR	d1599.p2p          .udpnet       	listen udp error, stopping server	{"node_id": "d1599f8566555276993fb7bb2f078d2b87d8e1c3879ddf26d386ba3c96b1e8a2", "module": "p2p", "p2pid": "83VdmNkSgdiJ63DqPyojr4Gu5ZUJ4SMn3a192U3LoM8D", "errmsg": "read udp [::]:7513: use of closed network connection", "name": "p2p.udpnet"}
2021-08-09T18:31:11.192+0300	INFO	d1599.timesync     	starting time sync round with peers	{"node_id": "d1599f8566555276993fb7bb2f078d2b87d8e1c3879ddf26d386ba3c96b1e8a2", "module": "timesync", "round": 12, "peers_count": 3, "errors_count": 0, "name": "timesync"}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x49d8fda]

goroutine 155 [running]:
github.com/spacemeshos/go-spacemesh/p2p.(*Switch).sendMessageImpl(0xc0003c6000, 0x4dbc118, 0xc005837800, 0x0, 0x0, 0x4c58e70, 0xe, 0x4da5300, 0xc005837860, 0x5161001, ...)
	/Users/nkryuchkov/GolandProjects/go-spacemesh/p2p/switch.go:351 +0x3a
github.com/spacemeshos/go-spacemesh/p2p.(*Switch).SendWrappedMessage(0xc0003c6000, 0x4dbc118, 0xc005837800, 0x0, 0x0, 0x4c58e70, 0xe, 0xc005837860, 0x4c4e72b, 0x3)
	/Users/nkryuchkov/GolandProjects/go-spacemesh/p2p/switch.go:336 +0x90
github.com/spacemeshos/go-spacemesh/p2p/server.(*MessageServer).SendRequest(0xc0001b32c0, 0x4dbc0e0, 0xc005837800, 0x7, 0xc00584c7c0, 0x8, 0x40, 0x0, 0x0, 0xc005837770, ...)
	/Users/nkryuchkov/GolandProjects/go-spacemesh/p2p/server/msgserver.go:278 +0x5ab
github.com/spacemeshos/go-spacemesh/timesync/peersync.(*Sync).GetOffset(0xc000232f00, 0x4dbc0e0, 0xc005166ea0, 0xc, 0xc005010630, 0x3, 0x3, 0x3, 0x0, 0xc0001c9980)
	/Users/nkryuchkov/GolandProjects/go-spacemesh/timesync/peersync/sync.go:291 +0x499
github.com/spacemeshos/go-spacemesh/timesync/peersync.(*Sync).run(0xc000232f00, 0x4dbc070, 0xc0002b4480, 0x0, 0x0)
	/Users/nkryuchkov/GolandProjects/go-spacemesh/timesync/peersync/sync.go:229 +0x585
github.com/spacemeshos/go-spacemesh/timesync/peersync.(*Sync).Start.func1.1(0x4dbc070, 0xc0002b4480, 0xc000032f60, 0xc000032f90)
	/Users/nkryuchkov/GolandProjects/go-spacemesh/timesync/peersync/sync.go:186 +0x3e
github.com/spacemeshos/go-spacemesh/taskgroup.(*Group).Go.func1(0xc000434dc0, 0xc000433630)
	/Users/nkryuchkov/GolandProjects/go-spacemesh/taskgroup/group.go:63 +0x71
created by github.com/spacemeshos/go-spacemesh/taskgroup.(*Group).Go
	/Users/nkryuchkov/GolandProjects/go-spacemesh/taskgroup/group.go:61 +0xea
```

### LINKED ISSUE #2654: TestShutdown sporadic failure
## Description
This test sporadically fails with hanging goroutines due to a race condition

## Additional Resources 
```
=== RUN   TestShutdown
2021-08-10T13:45:45.624-0400	WARN	00000.defaultLogger	attempt to SetLevel of logger lower than parent level, this may result in log entries being dropped silently
    app_test.go:748: goroutine 20 [running]:
        github.com/spacemeshos/go-spacemesh/cmd/node.TestShutdown(0xc000102900)
        	/home/user/go-spacemesh/cmd/node/app_test.go:746 +0xd88
        testing.tRunner(0xc000102900, 0x120f8c0)
        	/usr/lib/go-1.15/src/testing/testing.go:1123 +0xef
        created by testing.(*T).Run
        	/usr/lib/go-1.15/src/testing/testing.go:1168 +0x2b3
        
        goroutine 1 [chan receive]:
        testing.(*T).Run(0xc000102900, 0x11c045c, 0xc, 0x120f8c0, 0x49bc01)
        	/usr/lib/go-1.15/src/testing/testing.go:1169 +0x2da
        testing.runTests.func1(0xc000102780)
        	/usr/lib/go-1.15/src/testing/testing.go:1439 +0x78
        testing.tRunner(0xc000102780, 0xc00023fde0)
        	/usr/lib/go-1.15/src/testing/testing.go:1123 +0xef
        testing.runTests(0xc000117c20, 0x19afd40, 0xe, 0xe, 0x0, 0x0, 0x0, 0x410070)
        	/usr/lib/go-1.15/src/testing/testing.go:1437 +0x2fe
        testing.(*M).Run(0xc000120800, 0x0)
        	/usr/lib/go-1.15/src/testing/testing.go:1345 +0x1eb
        main.main()
        	_testmain.go:69 +0x138
        
        goroutine 148 [select]:
        github.com/spacemeshos/go-spacemesh/p2p/peers.(*Peers).listenToPeers(0xc000380330, 0xc00010fa40, 0xc00010faa0)
        	/home/user/go-spacemesh/p2p/peers/peers.go:74 +0x2d6
        created by github.com/spacemeshos/go-spacemesh/p2p/peers.NewPeers
        	/home/user/go-spacemesh/p2p/peers/peers.go:33 +0x16e
        
        goroutine 150 [select]:
        github.com/spacemeshos/go-spacemesh/p2p/peers.(*Peers).listenToPeers(0xc0003805d0, 0xc00010fce0, 0xc00010fd40)
        	/home/user/go-spacemesh/p2p/peers/peers.go:74 +0x2d6
        created by github.com/spacemeshos/go-spacemesh/p2p/peers.NewPeers
        	/home/user/go-spacemesh/p2p/peers/peers.go:33 +0x16e
        
    app_test.go:750: 
        	Error Trace:	app_test.go:750
        	Error:      	Not equal: 
        	            	expected: 2
        	            	actual  : 4
        	Test:       	TestShutdown
--- FAIL: TestShutdown (10.44s)


Expected :2
Actual   :4
<Click to see difference>


FAIL

Process finished with the exit code 1
```

### PR DESCRIPTION
**Pull request by dshulyak**
## Motivation
closes: https://github.com/spacemeshos/go-spacemesh/issues/2640
closes: https://github.com/spacemeshos/go-spacemesh/issues/2648
closes: https://github.com/spacemeshos/go-spacemesh/issues/2654

## Changes
- Moved WaitPeers functionality from timesync/peersync to p2p/peers as agreed in original PR
- Refactored Peers to call events.ReportNodeStatusUpdate only in an instance of Peers that is initialized in p2p/switch
- Refactored tests to avoid any time.Sleep

## Test Plan
unit tests

## TODO
- [x] Explain motivation or link existing issue(s)
- [x] Test changes and document test plan
- [ ] Update documentation as needed

## DevOps Notes

- [x] This PR does not require configuration changes (e.g., environment variables, GitHub secrets, VM resources)
- [x] This PR does not affect public APIs
- [x] This PR does not rely on a new version of external services (PoET, elasticsearch, etc.)
- [x] This PR does not make changes to log messages (which monitoring infrastructure may rely on)

### CODE REVIEW

**Comment by dshulyak** [comment]
bors try

File: timesync/peersync/sync.go [hunk 1]
```diff
@@ -182,21 +178,23 @@ func (s *Sync) requestHandler(ctx context.Context, buf []byte) []byte {
 // Start background workers.
 func (s *Sync) Start() {
 	s.once.Do(func() {
+		// NOTE(dshulyal) we can't start listening for peers in New because Simular
```
Comment (nkryuchkov) [inline timesync/peersync/sync.go]: `Simular` looks like a typo. What is it?

**Review by nkryuchkov — APPROVED** [review]
I tested it a few times and wasn't able to reproduce a panic

File: timesync/peersync/sync.go (hunk 1 above)
Comment (dshulyak) [reply timesync/peersync/sync.go]: it was Simulator

**Comment by dshulyak** [comment]
bors merge

**Pull request closed without merging**